### PR TITLE
Feature smart camera

### PR DESCRIPTION
**Pull request by DavidMansolino**
Linked with omichel/webots#5047

https://www.cyberbotics.com/doc/automobile/index?version=feature-smart-camera
https://www.cyberbotics.com/doc/reference/index?version=feature-smart-camera

### CODE REVIEW

**Comment by DavidMansolino** [comment]
:+1:

**Comment by DavidMansolino** [comment]
:+1: 

(Test is failing because Cyberbotics website is in maintenance, but I will anyway restart the test before merging).

**Comment by omichel** [comment]
:eyes: (cyberbotics web site is up-and-running again)

File: automobile/ackermannvehicle.md [hunk 1]
```diff
@@ -60,6 +62,8 @@ sensors, shape of the vehicle, etc.).
 - `boundingObject`: Physical geometry of the vehicle.
 - `physics`: `Physics` node of the vehicle defining the physical parameters of the
 vehicle.
+- `radarCrossSection`: Defines the `radarCrossSection` of the vehicle, if greater than 0, this vehicle is a potential target for any Radar device. It measures the vehicle ability to reflect radar signals in the direction of the radar receiver.
+- `recognitionColors`: Defines the `recognitionColors` of the vehicle. If not empty, this vehicle may be recognized by any Camera device with recognition capability.
```
Comment (omichel) [inline automobile/ackermannvehicle.md]: "Defines the `fieldName` of the vehicle" is useless and should be remove (as it doesn't explain anything more than the field name itself). Also, we should refer to the corresponding nodes, e.g., Radar and Recognition with hyperlinks.

File: automobile/camera-sensors.md [hunk 2]
```diff
@@ -1,7 +1,18 @@
...
+No specific model of camera have been modeled, but it is easy to create your
 own model using the Camera node of Webots which includes by default a zoom and a
 focusing mechanism. The Camera node also includes a model of motion blur,
 various models of noise, lens distortion and the possibility to use a spherical
 projection.
+
+
+### Smart camera
+
+If a `Recognition` node is added to the camera, the camera returns not only the image but also all the objects detected in the camera image and their size and position. It is therefore very easy to simulate smart cameras such as Mobileye one.
```
Comment (omichel) [inline automobile/camera-sensors.md]: I would say "can return" instead of "returns".

File: reference/camera.md [hunk 3]
```diff
@@ -520,3 +527,64 @@ images, the `quality` parameter is ignored.
...
+ double   size[2];
+ int      position_on_image[2];
+ int      size_on_image[2];
+ int      number_of_colors;
+ double  *colors;
+ char    *model;
+} WbCameraRecognitionObject;
+```
+
+The `id` represents the node id corresponding to the object, it is possible to use this id directly in the [wb_supervisor_node_get_from_id](supervisor.md#wb_supervisor_node_get_from_def) supervisor function. The `position` and `orientation` are expressed relatively to the camera (the relative position is the one of the center of the object which can differ from its origin) and the units are meters. The `size` represents the X and Y sizes in meters relatively to the camera (it is of course impossible to know the depth of the object). The `position_on_image` and `size_on_image` can be used to determine the bounding box of the object in the camera image, the units are pixels. The `number_of_colors` and `colors` returns respectively the number of colors of the objects and pointer to the colors array, each colors is represented by 3 doubles (R, G and B), therefore the size of the array is equal to 3 * `number_of_colors`. Finally `model` return the `model` field of the [Solid](solid.md) node.
```
Comment (omichel) [inline reference/camera.md]: ... and the units are **meter and radian**.
`model` return**s** the `model`.

File: reference/recognition.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+The [Recognition](#recognition) node provides a [Camera](camera.md) device with object recognition capability.
+When a [Camera](camera.md) device has a [Recognition](#recognition) node in its `recognition` field, it is able to recognize which objects are present in the camera image. Only [Solids](solid.md) whose `recognitionColors` field is not empty can be recognized by the camera.
+
+### Field Summary
+
+- The `maxRange` field defines the maximum distance at which an object can be recognized. Objects farther than `maxRange` are not recognized.
+
+- The `maxObjects` field defines the maximum number of objects detected by the camera. `-1` means no limit. If more objects are visible by the camera, only the `maxObjects` biggest ones (considering pixel size) are recognized.
+
+- The `occlusion` field defines if occlusions between the camera and the object should be checked. Disabling the occlusions detection can save computational time and therefore speed up the simulation, but it can lead to recognized object that are not really visible to the camera.
```
Comment (omichel) [inline reference/recognition.md]: As you told me the impact on the simulation speed is low. What is interesting in disabling occlusion is that recognition can work through thin or transparent objects. We should probably explain this.

File: automobile/ackermannvehicle.md (hunk 1 above)
Comment (DavidMansolino) [inline automobile/ackermannvehicle.md]: We usually don't use hyperlinks from one book to another, I will try to see if it works fine.

Comment (DavidMansolino) [inline automobile/ackermannvehicle.md]: Unfortunately it seems to work on Github but not in our viewer (on Cyberboitics website).

Comment (omichel) [inline automobile/ackermannvehicle.md]: It's a pity as I believe it could be useful to make such cross-references... Can't we fix that?

Comment (DavidMansolino) [inline automobile/ackermannvehicle.md]: I just opened an issue about it.

Comment (omichel) [inline automobile/ackermannvehicle.md]: Great. Meanwhile, we can mention in this this new issue that those `radarCrossSection` and `recognitionColors` section could be hyperlinked to the reference manual, so that we can fix it once it work and leave it like that for now without hyperlink.

**Comment by DavidMansolino** [comment]
:+1:

**Comment by omichel** [comment]
:eyes:

File: automobile/ackermannvehicle.md [hunk 5]
```diff
@@ -62,8 +62,8 @@ sensors, shape of the vehicle, etc.).
 - `boundingObject`: Physical geometry of the vehicle.
 - `physics`: `Physics` node of the vehicle defining the physical parameters of the
 vehicle.
-- `radarCrossSection`: Defines the `radarCrossSection` of the vehicle, if greater than 0, this vehicle is a potential target for any Radar device. It measures the vehicle ability to reflect radar signals in the direction of the radar receiver.
-- `recognitionColors`: Defines the `recognitionColors` of the vehicle. If not empty, this vehicle may be recognized by any Camera device with recognition capability (i.e. with a Recognition recognition node).
+- `radarCrossSection`: If greater than 0, this vehicle is a potential target for any Radar device. It measures the vehicle ability to reflect radar signals in the direction of the radar receiver.
+- `recognitionColors`: If not empty, this vehicle may be recognized by any Camera device with recognition capability (i.e. with a Recognition recognition node).
```
Comment (omichel) [inline automobile/ackermannvehicle.md]: I would remove the second "recognition" word here to keep simple "with a Recognition node)."

Comment (DavidMansolino) [inline automobile/ackermannvehicle.md]: You are completely right, I did a mistake when I removed the non-working link.

File: automobile/ackermannvehicle.md [hunk 6]
```diff
@@ -60,6 +62,8 @@ sensors, shape of the vehicle, etc.).
 - `boundingObject`: Physical geometry of the vehicle.
 - `physics`: `Physics` node of the vehicle defining the physical parameters of the
 vehicle.
+- `radarCrossSection`: If greater than 0, this vehicle is a potential target for any Radar device. It measures the vehicle ability to reflect radar signals in the direction of the radar receiver.
```
Comment (tn12787) [inline automobile/ackermannvehicle.md]: vehicle ability -> vehicle's ability

File: automobile/ackermannvehicle.md [hunk 7]
```diff
@@ -60,6 +62,8 @@ sensors, shape of the vehicle, etc.).
...
+- `recognitionColors`: If not empty, this vehicle may be recognized by any Camera device with recognition capability (i.e. with a Recognition recognition node).
```
Comment (tn12787) [inline automobile/ackermannvehicle.md]: I wouldn't repeat recognition here, it seems tautological. Maybe rephrase?

File: automobile/camera-sensors.md [hunk 8]
```diff
@@ -1,7 +1,18 @@
...
+No specific model of camera have been modeled, but it is easy to create your
 own model using the Camera node of Webots which includes by default a zoom and a
 focusing mechanism. The Camera node also includes a model of motion blur,
 various models of noise, lens distortion and the possibility to use a spherical
 projection.
+
+
+### Smart camera
+
+If a `Recognition` node is added to the camera, the camera can return not only the image but also all the objects detected in the camera image and their size and position. It is therefore very easy to simulate smart cameras such as Mobileye one.
```
Comment (tn12787) [inline automobile/camera-sensors.md]: You should remove the "the" before "objects".

Comment (tn12787) [inline automobile/camera-sensors.md]: "smart cameras such as Mobileye one." -> "smart cameras such as the Mobileye"

File: reference/camera.md [hunk 9]
```diff
@@ -106,6 +107,12 @@ no focus is available on the camera device.
 with a controllable zoom system. If this field is set to NULL, then no zoom is
 available on the camera device.
 
+- The `recognition` field may contain a [Recognition](recognition.md) node to provide the camera device with object recognition capability.
```
Comment (tn12787) [inline reference/camera.md]: capability -> capabilities

File: reference/camera.md [hunk 10]
```diff
@@ -106,6 +107,12 @@ no focus is available on the camera device.
...
+The camera can only recognize [Solid](solid.md) nodes whose `recognitionColors` is not empty.
+The object size is estimated using the `boundingObject` of the [Solid](solid.md) and the ones of all its children.
+Note that the returned size is an estimation and can be in some cases overestimated.
+In case the [Solid](solid.md) and its children doesn't have any bounding object, the dimension is estimated using the shape, this estimation is usually widely overestimated.
```
Comment (tn12787) [inline reference/camera.md]: "and its children" needs parentheses if you wish to use doesn't instead of don't here:

In case the [Solid](solid.md) (and its children) doesn't have..

OR

In case the [Solid](solid.md) and its children don't have..

File: reference/camera.md [hunk 11]
```diff
@@ -520,3 +527,64 @@ images, the `quality` parameter is ignored.
...
+const WbCameraRecognitionObject *wb_camera_recognition_get_objects(WbDeviceTag tag);
+```
+
+**Description**
+
+If a [Recognition](recognition.md) node is present in the `recognition` field, the camera can recognize objects in its image.
+
+The `wb_camera_has_recognition` function can be used to determine whether a [Recognition](recognition.md) node is present or not.
+
+The `wb_camera_recognition_enable()` function allows the user to enable the recognition, note that it is not needed to enable the camera to let recognition work.
```
Comment (tn12787) [inline reference/camera.md]: You can also remove the "the" before "recognition".  

"enable recognition, note that it is not needed to enable the camera to let recognition work." -> "..enable recognition. It is not necessary to enable the camera for recognition to work."

File: reference/camera.md [hunk 12]
```diff
@@ -520,3 +527,64 @@ images, the `quality` parameter is ignored.
...
+ double   size[2];
+ int      position_on_image[2];
+ int      size_on_image[2];
+ int      number_of_colors;
+ double  *colors;
+ char    *model;
+} WbCameraRecognitionObject;
+```
+
+The `id` represents the node id corresponding to the object, it is possible to use this id directly in the [wb_supervisor_node_get_from_id](supervisor.md#wb_supervisor_node_get_from_def) supervisor function. The `position` and `orientation` are expressed relatively to the camera (the relative position is the one of the center of the object which can differ from its origin) and the units are meter and radian. The `size` represents the X and Y sizes in meters relatively to the camera (it is of course impossible to know the depth of the object). The `position_on_image` and `size_on_image` can be used to determine the bounding box of the object in the camera image, the units are pixels. The `number_of_colors` and `colors` returns respectively the number of colors of the objects and pointer to the colors array, each colors is represented by 3 doubles (R, G and B), therefore the size of the array is equal to 3 * `number_of_colors`. Finally `model` returns the `model` field of the [Solid](solid.md) node.
```
Comment (tn12787) [inline reference/camera.md]: I would add "and" before "it is possible" on the first line.

"each colors" -> "each color"

File: reference/recognition.md [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+### Description
+
+The [Recognition](#recognition) node provides a [Camera](camera.md) device with object recognition capability.
+When a [Camera](camera.md) device has a [Recognition](#recognition) node in its `recognition` field, it is able to recognize which objects are present in the camera image. Only [Solids](solid.md) whose `recognitionColors` field is not empty can be recognized by the camera.
+
+### Field Summary
+
+- The `maxRange` field defines the maximum distance at which an object can be recognized. Objects farther than `maxRange` are not recognized.
+
+- The `maxObjects` field defines the maximum number of objects detected by the camera. `-1` means no limit. If more objects are visible by the camera, only the `maxObjects` biggest ones (considering pixel size) are recognized.
```
Comment (tn12787) [inline reference/recognition.md]: "visible by the camera" -> "visible to the camera"

File: reference/recognition.md [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+- The `occlusion` field defines if occlusions between the camera and the object should be checked. Disabling the occlusion can be useful to allow the camera to see through thin or transparent objects that may hide the center of the object we are interested in, but it can lead to recognized object that are not really visible to the camera. Additionally, it will slightly speed up the simulation.
```
Comment (tn12787) [inline reference/recognition.md]: "recognized object" -> "recognized objects"

File: reference/camera.md (hunk 10 above)
Comment (tn12787) [inline reference/camera.md]: See edits, the nouns actually have to agree too.

**Comment by DavidMansolino** [comment]
:+1:

File: reference/recognition.md [hunk 15]
```diff
@@ -0,0 +1,28 @@
+## Recognition
+
+```
+Recognition {
+  SFFloat  maxRange       100     # maximum distance at which objects are recognized (m)
+  SFInt32  maxObjects     -1      # maximum number of objects detected simultaneously
+  SFBool   occlusion      TRUE    # should occlusions between the camera and the objects be checked
+  SFColor  frameColor     1 0 0   # color of the objects frame in the overlay
```
Comment (tn12787) [inline reference/recognition.md]: sorry, just seen this - "objects" -> "object's"

File: reference/recognition.md [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+  SFInt32  frameThickness 1       # thickness of the objects frame in the overlay (pixels)
```
Comment (tn12787) [inline reference/recognition.md]: same here.

**Comment by DavidMansolino** [comment]
:+1:
